In [2]:
import pandas as pd
import numpy as np
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.preprocessing import LabelEncoder
from scipy.sparse import coo_matrix



pd.set_option('display.max_colwidth', None)

In [31]:
basepath = '../../datasets/preprocessed_datasets/gabor/'
data = pd.read_pickle(basepath +'user_item_interactions.pkl')
data

customer_id  product_id  amount
0           1940761     4839681       1
1           1940761     4839682       1
2           1940761     6523930       1
3           1940761     6596395       1
4           1940761     8642222       1
...             ...         ...     ...
644783     11128024    10560958       1
644784     11128039    10451840       1
644785     11128082    10559336       1
644786     11128095    10504300       1
644787     11128129    10366497       1

[644788 rows x 3 columns]

In [33]:
product_data = pd.read_pickle(basepath +'only_products.pkl')
product_data[['product__id','productName',  'gender', 'line', 'b2cPrice']]

product__id                                       productName gender  \
0            557559          sportliche Ballerinas Glattleder schwarz      w   
30           551622                 elegante Pumps Glattleder schwarz      w   
108          547193                        Slipper Glattleder schwarz      w   
205          549776                         Sneaker low Rauleder blau      w   
241          550599                         Sneaker low Rauleder blau      w   
...             ...                                               ...    ...   
648486     10587121                            Slipper Rauleder beige      w   
648487     10366497                Sneaker low Materialmix Leder weiß      w   
648489     10227216  PG8000 Sneaker low Materialmix Leder/Textil grün      m   
648490     10230388                    PG0460 Sneaker low Textil blau      m   
648491     10227507    Sneaker low Materialmix Leder/Lederimitat rosa      w   

           line  b2cPrice  
0       F-S|H-W     99.95  
30      F-S|H-W     89.95  
108     F-S|H-W     79.95  
205     F-S|H-W    140.00  
241     F-S|H-W    120.00  
...         ...       ...  
648486      F-S     99.95  
648487      F-S    140.00  
648489      F-S    140.00  
648490      F-S     99.95  
648491      F-S     99.95  

[108777 rows x 5 columns]

In [4]:
encoder = LabelEncoder()
data['customer_id_encoded'] = encoder.fit_transform(data['customer_id'])
data['product_id_encoded'] = encoder.fit_transform(data['product_id'])

# Create a user-item matrix
user_item_matrix = coo_matrix((data['amount'], (data['customer_id_encoded'], data['product_id_encoded'])))
user_item_matrix = user_item_matrix.tocsr()

# Calculate the cosine similarity between products
product_similarity = cosine_similarity(user_item_matrix.T)


In [28]:
def recommend_products(user_id, n=5):
    user_ratings = user_item_matrix[user_id].toarray().flatten()
    similar_products = product_similarity[user_ratings == 0].sum(axis=0)
    recommended_product_indices = similar_products.argsort()[-n:][::-1]
    recommended_products = encoder.inverse_transform(recommended_product_indices)
    recommended_scores = similar_products[recommended_product_indices]
    return recommended_products, recommended_scores

def get_products_from_customer(cust_id):
    return data[data['customer_id_encoded'] == cust_id]

# Example: Recommend products for a specific user
user_id_to_recommend = 7963  # Change this to the desired user ID
recommended_products, recommended_scores = recommend_products(user_id_to_recommend)


In [29]:
print(f"Recommended products for user {user_id_to_recommend}:")
for i, (product_id, score) in enumerate(zip(recommended_products, recommended_scores)):
    product = product_data[product_data['product__id'] == product_id]
    print(f"{i+1}. Product ID: {product_id}, Score: {score:.4f}, product name: {product.productName}")


Recommended products for user 7963:
1. Product ID: 4085521, Score: 43.7362, product name: 104926    Sneaker high Materialmix Leder braun
Name: productName, dtype: object
2. Product ID: 3711672, Score: 41.9317, product name: 126239    Plateau Pump Leder Reptiloptik schwarz
Name: productName, dtype: object
3. Product ID: 4776107, Score: 41.9317, product name: 274386    Keilpumps Rauleder blau
Name: productName, dtype: object
4. Product ID: 3711325, Score: 41.9317, product name: 162031    Slipper Glattleder schwarz
Name: productName, dtype: object
5. Product ID: 6554512, Score: 41.9317, product name: 370124    Slipper Rauleder blau
Name: productName, dtype: object


In [30]:

customer_products = get_products_from_customer(user_id_to_recommend)
customer_products_data = product_data[product_data['product__id'].isin(customer_products.product_id)]
display(customer_products_data)

product__id  productNumber                  productName productSizeEu  \
253599      5337332  4064032019207  Pantolette Glattleder beige          38.5   

       productColorName articleNumber  brand__id  mainCategory__id gender  \
253599            beige     62.790.52    6590678            315572      w   

       originCountry  ... shoeStyle  productType_y heelHeightGroup  sizeEu  \
253599            PT  ...  Sandalen    Damenschuhe        bis 3 cm    38.5   

        sizeUk  sizeIndex  shaftLength__value  sole isTransferee isSuccessor  
253599     5.5        9.0                 0.0   TPU          0.0         0.0  

[1 rows x 55 columns]

Recommended products for user 1000:
1. Product ID: 4085521, Score: 43.7362, product name: 104926    Sneaker high Materialmix Leder braun
Name: productName, dtype: object
2. Product ID: 3711325, Score: 41.9317, product name: 162031    Slipper Glattleder schwarz
Name: productName, dtype: object
3. Product ID: 3711672, Score: 41.9317, product name: 126239    Plateau Pump Leder Reptiloptik schwarz
Name: productName, dtype: object
4. Product ID: 4776107, Score: 41.9317, product name: 274386    Keilpumps Rauleder blau
Name: productName, dtype: object
5. Product ID: 5507440, Score: 41.9317, product name: 283005    Pantolette Rauleder rot
Name: productName, dtype: object


Recommended products for user 345:
1. Product ID: 4085521, Score: 43.7362, product name: 104926    Sneaker high Materialmix Leder braun
Name: productName, dtype: object
2. Product ID: 6113537, Score: 41.9317, product name: 364202    Plateau Stiefelette Rauleder braun
Name: productName, dtype: object
3. Product ID: 3711672, Score: 41.9317, product name: 126239    Plateau Pump Leder Reptiloptik schwarz
Name: productName, dtype: object
4. Product ID: 3711325, Score: 41.9317, product name: 162031    Slipper Glattleder schwarz
Name: productName, dtype: object
5. Product ID: 5507440, Score: 41.9317, product name: 283005    Pantolette Rauleder rot
Name: productName, dtype: object


Recommended products for user 7963:
1. Product ID: 4085521, Score: 43.7362, product name: 104926    Sneaker high Materialmix Leder braun
Name: productName, dtype: object
2. Product ID: 3711672, Score: 41.9317, product name: 126239    Plateau Pump Leder Reptiloptik schwarz
Name: productName, dtype: object
3. Product ID: 4776107, Score: 41.9317, product name: 274386    Keilpumps Rauleder blau
Name: productName, dtype: object
4. Product ID: 3711325, Score: 41.9317, product name: 162031    Slipper Glattleder schwarz
Name: productName, dtype: object
5. Product ID: 6554512, Score: 41.9317, product name: 370124    Slipper Rauleder blau
Name: productName, dtype: object
